Sync/StorageClient.swift

```diff
@@ -605,6 +622,80 @@ public class Sync15StorageClient {
     }
 }
 
+public class Sync15BatchClient<T: CleartextPayloadJSON> {
+    let batchToken: BatchToken
+    private let collectionURI: NSURL
+    private let postClient: POSTClient15<T>
+
+    init(batchToken: BatchToken, collectionURI: NSURL, postClient: POSTClient15<T>) {
+        self.batchToken = batchToken
+        self.collectionURI = collectionURI
+        self.postClient = postClient
+    }
+
+    func uploadRecords(lines: [String], ifUnmodifiedSince: Timestamp?) -> Deferred<Maybe<StorageResponse<POSTResult>>> {
+        let batchURI = collectionURI.URLByAppendingPathComponent("batch=\(batchToken)")
+        return postClient.post(batchURI, lines: lines, ifUnmodifiedSince: ifUnmodifiedSince)
+    }
+
+    func uploadRecords(records: [Record<T>], ifUnmodifiedSince: Timestamp?) -> Deferred<Maybe<StorageResponse<POSTResult>>> {
+        let lines = optFilter(records.map(postClient.serializeRecord))
+        return uploadRecords(lines, ifUnmodifiedSince: ifUnmodifiedSince)
+    }
+
+    func commit() -> Deferred<Maybe<StorageResponse<POSTResult>>> {
+        let commitURI = collectionURI.URLByAppendingPathComponent("batch=\(batchToken)").URLByAppendingPathComponent("commit=true")
+        return postClient.post(commitURI, records: [], ifUnmodifiedSince: nil)
+    }
+}
+
+public class POSTClient15<T: CleartextPayloadJSON> {
+    private let queue: dispatch_queue_t
+    private let client: Sync15StorageClient
+    private let encrypter: RecordEncrypter<T>
+
+    public init(queue: dispatch_queue_t, storageClient: Sync15StorageClient, encrypter: RecordEncrypter<T>) {
+        self.queue = queue
+        self.client = storageClient
+        self.encrypter = encrypter
+    }
+
+    public func post(uri: NSURL, records: [Record<T>], ifUnmodifiedSince: Timestamp?) -> Deferred<Maybe<StorageResponse<POSTResult>>> {
+        // TODO: charset
+        // TODO: if any of these fail, we should do _something_. Right now we just ignore them.
+        let lines = optFilter(records.map(self.serializeRecord))
+        return self.post(uri, lines: lines, ifUnmodifiedSince: ifUnmodifiedSince)
+    }
+
+    // Exposed so we can batch by size.
+    public func serializeRecord(record: Record<T>) -> String? {
+        return self.encrypter.serializer(record)?.toString(false)
+    }
+
+    public func post(uri: NSURL, lines: [String], ifUnmodifiedSince: Timestamp?) -> Deferred<Maybe<StorageResponse<POSTResult>>> {
+        let deferred = Deferred<Maybe<StorageResponse<POSTResult>>>(defaultQueue: client.resultQueue)
```

Comment:
I'm not sure if we're able, thanks to generic protocol limitations, but it would be nice if both `Sync15CollectionClient` and `Sync15BatchClient` implemented the same protocol.